docs/guide/porting_mechanisms_to_cpp.rst

```diff
@@ -0,0 +1,259 @@
+overload.
+
+K&R function definitions
+^^^^^^^^^^^^^^^^^^^^^^^^
+C supports a legacy ("K&R") syntax for function declarations. This is slated
+for removal from the C language, and is not valid C++.
+
+There is no advantage to the legacy syntax. If you have legacy definitions such
+as
+
+.. code-block:: c
+
+  void foo(a, b) int a, b; { /* ... */ }
+
+then simply replace them with
+
+.. code-block:: cpp
+
+  void foo(int a, int b) { /* ... */ }
+
+which is valid in both C and C++.
+
+Legacy patterns that are considered deprecated
+----------------------------------------------
+As noted above (:ref:`local-non-prototype-function-declaration`), declarations
+such as
+
+.. code-block:: c
+
+  VERBATIM
+  extern void vector_resize();
+  ENDVERBATIM
+
+at the global scope in MOD files declare C++ function overloads that take no
+parameters.
+If such declarations appear at the global scope then they do not hide the
+correct declarations, so this can be harmless, but it is not necessary.
+In |neuron_with_cpp_mechanisms| the full declarations of the NEURON API methods
+that can be used in MOD files are implicitly included via the ``mech_api.h``
+header, so this explicit declaration of a zero-parameter overload is not needed
+and can safely be removed.
+
+.. _deprecated-overloads-taking-void:
+
+Deprecated overloads taking ``void*``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+As noted above (:ref:`function-decls-with-incorrect-types`),
+|neuron_with_cpp_mechanisms| provides extra overloads for some API methods that
+aim to smooth the transition from C to C++.
+These overloads are provided for widely used methods, and where overloading on
+return type would not be required.
+
+An example is the ``vector_capacity`` function, which in
+|neuron_with_cpp_mechanisms| has two overloads
+
+.. code-block:: cpp
+
+  int vector_capacity(IvocVect*);
+  [[deprecated("non-void* overloads are preferred")]] int vector_capacity(void*);
+
+The second one simply casts its argument to ``IvocVect*`` and calls the first
+one.
+The ``[[deprecated]]`` attribute means that MOD files that use the second
+overload emit compilation warnings when they are compiled using ``nrnivmodl``.
+
+If your MOD files produce these deprecation warnings, make sure that the
+relevant method (``vector_capacity`` in this example) is being called with an
+argument of the correct type (``IvocVect*``), and not a type that is implicitly
+converted to ``void*``.
+
+Automatic transformations for C++ Compatibility
+-----------------------------------------------
+
+Some common issues in the MOD file for C++ compatibility can be identified and
+translated automatically, by applying regular-expression-based transformations
+to ``VERBATIM`` blocks in MOD files.
+A relevant set of transformations can be enabled by passing the
+``-legacytransformations cxx`` option to ``nrnivmodl`` when compiling your MOD
+files:
+
+.. code-block:: bash
+
+    nrnivmodl -legacytransformations cxx mod-dir
+
+
```

Comment:
@olupton : do we want this to be visible to users? I assume not...

Reply:
Shouldn't we be dropping ` -legacytransformations` anyways?

Reply:
Yes, let's drop that and update the documentation. I can clean this up tomorrow.